_studio/shared/umc/codec/mpeg2_dec/include/umc_mpeg2_decoder.h

```diff
@@ -0,0 +1,245 @@
+
+        // Decode a bitstream header NAL unit
+        virtual UMC::Status DecodeHeaders(RawUnit & unit);
+
+        virtual MPEG2DecoderFrame* FindFrameByMemID(UMC::FrameMemID);
+
+        // Find previous (in decoded order) reference frame
+        virtual MPEG2DecoderFrame* FindPrevRefFrame(MPEG2DecoderFrame const*);
+
+        // Find a next frame ready to be output from decoder
+        virtual MPEG2DecoderFrame* GetFrameToDisplay();
+
+        // Check for frame completeness and get decoding errors
+        virtual bool QueryFrames(MPEG2DecoderFrame&) = 0;
+
+        // Initialize mfxVideoParam structure based on decoded bitstream header values
+        virtual UMC::Status FillVideoParam(mfxVideoParam *par, bool full);
+
+        // Return raw sequence and sequence extension headers
+        virtual RawHeader_MPEG2 * GetSeqAndSeqExtHdr()
+        { return &m_rawHeaders; }
+
+        // Return a number of cached frames
+        virtual uint8_t GetNumCachedFrames();
+
+        // Set frame display time
+        virtual void PostProcessDisplayFrame(MPEG2DecoderFrame*);
+
+        virtual void SetVideoParams(const mfxVideoParam&);
+
+        Payload_Storer * GetPayloadStorer() const { return m_messages.get();}
+
+    protected:
+
+        virtual void SetDPBSize(uint32_t);
+        virtual MPEG2DecoderFrame* GetFreeFrame();
+        virtual MPEG2DecoderFrame* GetFrameBuffer(MPEG2Slice*);
+
+        virtual UMC::Status CompletePicture(MPEG2DecoderFrame&, uint8_t);
+        virtual void UpdateDPB(MPEG2DecoderFrame&, uint8_t);
+
+        virtual void AllocateFrameData(UMC::VideoDataInfo const&, UMC::FrameMemID, MPEG2DecoderFrame&) = 0;
+        virtual void CompleteDecodedFrames();
+        virtual UMC::Status Submit(MPEG2DecoderFrame&, uint8_t) = 0;
+
+    private:
+
+        template <typename F>
+        MPEG2DecoderFrame* FindFrame(F pred);
+
+        // Find a reusable frame initialize it with slice parameters
+        UMC::Status StartFrame(MPEG2Slice * slice);
+
+    protected:
+
+        std::mutex                      m_guard;
+        UMC::FrameAllocator*            m_allocator;
+        uint32_t                        m_counter;
+        MPEG2DecoderParams              m_params;
+        mfxVideoParam                   m_firstMfxVideoParams;
+
+        DPBType                         m_dpb;     // store of decoded frames
+
+        MPEG2DecoderFrame*              m_currFrame;
+
+        struct Headers
+        {
+            std::shared_ptr<MPEG2SequenceHeader>           seqHdr;
+            std::shared_ptr<MPEG2SequenceExtension>        seqExtHdr;
+            std::shared_ptr<MPEG2SequenceDisplayExtension> displayExtHdr;
+            std::shared_ptr<MPEG2PictureHeader>            picHdr;
+            std::shared_ptr<MPEG2PictureCodingExtension>   picExtHdr;
+            std::shared_ptr<MPEG2QuantMatrix>              qmatrix;
+        };
+
+        Headers                         m_currHeaders;
```

Comment:
Where do we use _m_currHeaders_? Why do we need headers to be shared?

Reply:
Used for slice initialization. 
At the time of slice decoding, decoder may have headers different from the headers used for previous slice initialization.